system.fn.diff

```diff
@@ -0,0 +1,35 @@
+--- /www/rega/esp/system.fn.orig
++++ /www/rega/esp/system.fn
+@@ -859,6 +859,32 @@
+     svValue="";
+   }
+ 
++string SingleCond; object SingleCondObjekt;
++foreach(SingleCond,dom.GetObject(ID_SCONDITIONS).EnumIDs() ){
++	SingleCondObjekt = dom.GetObject(SingleCond);
++    if (SingleCondObjekt) {
++    	if (((SingleCondObjekt.LeftVal() == varid) && (SingleCondObjekt.LeftValType()== ivtSystemId ) )) {
```

Comment:
What's `varid` here?

Reply:
Germany, please.

Also in den Zeilen darüber ist varid noch definiert, warum sollte es dann hier nicht mehr sein?

Reply:
OK, ich seh gerade, der Wert wird als
system.GetVar("varid")
angesprochen.
Hat im test aber funktioniert.Ich schau mir das morgen nochmal an.

Was ich beim PR wieder falsch gemacht habe, habe ich aber noch nicht verstanden.

Reply:
Wo denn bitte? Ich seh nur folgendes:

https://github.com/jens-maus/RaspberryMatic/blob/4e2ea4062c08e634cc0705357640f761ecb30c25/buildroot-external/patches/occu/0097-WebUI-SetSysvarValueByUser/occu/WebUI/www/rega/esp/system.fn#L785

Und das ist ja keine variable, das ist eine Zeichenkette die referenziert wird soweit ich das überblicke.

Und bitte nur PRs die auch so getestet wurden. Sieht für mich so aus als ob das nicht wirklich getestet wurde sondern blind so erstellt wurde. Und das system.fn.diff ins top-level dir des repos abzulegen leider auch nicht richtig. Diese Änderung gehört in einen WebUI Patch integriert.

Reply:
> Was ich beim PR wieder falsch gemacht habe, habe ich aber noch nicht verstanden.

Na, die frage ist doch, was soll bitte mit dieser `system.fn.diff` Datei passieren die du da einfach ins top-level directory des Repositories abgelegt hast? Die wird ja vom build system so nicht verwendet und somit landet deine änderung dann auch nicht im WebUI code.

Das ganze gehört eben in einen WebUI patch unterhalb von `buildroot-external/patches/occu/` gepackt genauso wie Jerome und ich das hier auch machen und dann gehört `create_patches.sh` aufgerufen um die `*.patch` Datei zu generieren. Müsstest du dir nur entsprechend abschauen wie das mit den WebUI patches so funktioniert bzw. deine Änderung z.B. auch in den `0097-WebUI-SetSysvarValueByUser` patchset einarbeiten.

Reply:
> Ich schau mir das morgen nochmal an.

Wenn du schon dabei bist, dann pass biete auch den Codestyle gleich entsprechend an die Umgebung an. D.h. in der Datei z.B. keine Tabs sondern spaces und zwei spaces per einrückung und auch bzgl. Klammerung, keine deutschen Variablenbegriffe usw. Tut mir leid wenn das etwas strikt rüberkommt, aber ich muss da ein gewisses Prozedere auferlegen damit das überhaupt chancen hat irgendwann vielleicht mal in der CCU3 Firmware aufgenommen zu werden und die Codequalität auch hoch bleibt.

Reply:
Ja, die EQ3 Code Qualität ist in der Tat ein Maßstab.

Mal ganz im Ernst : Du beklagst regelmäßig das sich zu wenig Leute beteiligen. Wenn sich dann doch mal jemand beteiligt, dann erhältst du nur, was man alles falsch macht. Ich bin ja echt nicht schnell beleidigt, aber du bzw Github machst es einem echt schwer.Ich habe mir den PR von Jerome genau angeschaut und versucht alles identisch zu machen. Aber schon wo man den Patch hinzu packen hat und wie das überhaupt geht konnte ich nicht erkennen. Mach doch bitte mal einen One-Pager fertig, wie man einen RaspberryMatic PR korrekt einreicht.

Reply:
> Codestyle

Und hast du vielleicht noch einen link auf den styling guide? Ich finde da einfach nichts.

Reply:
Ich versteh deine Argumente und dein teilweisen Frust und ich meine das alles nicht so hart wie es vllt. am Schluss bei dir ankommt. Und wenn dir das zuviel ist den PR zu überarbeiten und aus deinem einfachen diff file einen kompletten WebUI patch zu generieren, dann kannst du gerne das diff nochmal aktualisieren das er wirklich funktioniert und durchgetestet ist und dann bastel ich da den WebUI patch draus und dann kannst du dir am Schluss das gerne einfach nochmal anschauen. Und ja, ich kann auch die Entwicklerdoku nochmal anpassen und vllt. den Schritt der Generierung eines WebUI patches dort erläutern, dauert aber halt etwas.

> Und hast du vielleicht noch einen link auf den styling guide? Ich finde da einfach nichts.

Einen einheitlichen styling guide gibt es natürlich nicht, auch weil sich der coding style von WebUI Datei zu WebUI Datei mitunter stark unterscheidet. Aber da es hier um das generieren von Patches gegen upstream geht muss das generelle Herangehen sein, das man sich den aktuellen Style (z.B.. Tabs vs Spaces, UTF-8 vs. ISO kodiert, 2 spaces vs. 4 spaces indentation, Klammerungsart, usw.) in der Datei anschaut die man editiert und dann versucht genau so seine änderungen einzupflegen damit der Style der Datei konsistent bleibt.

Reply:
> Einen einheitlichen styling guide gibt es natürlich nicht

Also wenn ich das richtig verstehe sind die ganzen Design-Richtlinien nur in deinem Kopf. Nun, dann musst Du Dich natürlich nicht wundern, wenn das nicht sooo smooth läuft. Ohne das schriftlich zu fixirien kannst Du keine definierte Qualität sicherstellen. Und nicht jeder Programmierer möchte erstmal forensich alte Dateien untersuchen. Und da muss ich nochmal polemisch werden: wenn EQ3 den Code wegen 2 Leerzeichen zu viel ablehnt, dann hat deren QS den Fokus auf den falschen Dingen. Deine Hoffnung, das EQ3 jemals nennenswerte Teile von RM übernimmt teile ich sowieso nicht. Aber OK, ich will ja durchaus nach deinen Spielreglen spielen - aber dann bitte auch mal irgendwo niederlegen.

Reply:
> Das ganze gehört eben in einen WebUI patch unterhalb von buildroot-external/patches/occu/ gepackt 

OK, vestehe. Leider bin ich im Prozess der PR Erstellung nicht auf den Punkt gestoßen, wo ich das hätte festlegen können. Vielleicht hast Du da nochmal einen Tipp für mich, wo ich welchen Schritt vergessen habe.

> Müsstest du dir nur entsprechend abschauen wie das mit den WebUI patches so funktioniert bzw. deine Änderung z.B. auch in den 0097-WebUI-SetSysvarValueByUser patchset einarbeiten

Möchtest Du jetzt einen extra Patch haben (mit der nächsten frieen Nummer?) oder soll das "irgendwie" in den 0097 rein? Und wenn ja, wie geht das wieder?

> So please revise your PR if possible.

Ja, da habe ich auch mehr Fragen als Antworten. 
Wie ändert man einen PR? Änder ich den in meinem Repository oder in deinem? oder beiden? Oder besser löschen und neu anfangen?

Reply:
Und jetzt nochmal was zum Code. Vielleicht war das doch etwas zu kurz gesprungen. (obwohl er getestet wurde)

Wenn ich es richtig verstehe wird die function ::saveSysVar() sowohl für neu angelegte SV aufgerufen, als auch für geänderte. 
Das schließe ich aus der Zeile
https://github.com/jens-maus/RaspberryMatic/blob/4e2ea4062c08e634cc0705357640f761ecb30c25/buildroot-external/patches/occu/0097-WebUI-SetSysvarValueByUser/occu/WebUI/www/rega/esp/system.fn#L767

Dieser IF-Block endet aber in Zeile 806. Danach kommen - unabhängig ob neu oder geändert - der Block
https://github.com/jens-maus/RaspberryMatic/blob/4e2ea4062c08e634cc0705357640f761ecb30c25/buildroot-external/patches/occu/0097-WebUI-SetSysvarValueByUser/occu/WebUI/www/rega/esp/system.fn#L807-L850

Daran hatte ich mich orientiert. Das bedeutet aber auch, das mein Code immer ausgeführt wird - ob neue oder geänderte SV. Das ist bei einer neuen SV erstmal unschädlich, da die ja noch in keinem Programm ist. Aber vielleicht wäre es doch besser den Block nur auszuführen, wenn ( system.GetVar("createNew") == 0 ) ?
Was meinst Du? Oder gleich hier einfügen?
https://github.com/jens-maus/RaspberryMatic/blob/4e2ea4062c08e634cc0705357640f761ecb30c25/buildroot-external/patches/occu/0097-WebUI-SetSysvarValueByUser/occu/WebUI/www/rega/esp/system.fn#L776-L781

Wobei mich daran stören würde, das ich die Programme dann vor den SV ändere. 

Und eine Frage die ich noch testen muss: wird dieser Code nur beim Ändern des SV Typ ausgeführt oder auch bei Änderungen der Beschreibung, Name, etc? Denn dann hätte das ungewollte Effekte. Dann müsste ich erst noch testen, ob sich der Typ überhaupt geändert hat.

Reply:
> > Müsstest du dir nur entsprechend abschauen wie das mit den WebUI patches so funktioniert bzw. deine Änderung z.B. auch in den 0097-WebUI-SetSysvarValueByUser patchset einarbeiten
> 
> Möchtest Du jetzt einen extra Patch haben (mit der nächsten frieen Nummer?) oder soll das "irgendwie" in den 0097 rein? Und wenn ja, wie geht das wieder?

Ich seh schon, das ich wohl doch schneller als erwartet das Prozedere rund um WebUI Patches doch mal dokumentieren müsste damit andere und nicht nur "die Eingeweihten" diese erstellen können und auch die Logik des ganzen nachvollziehen können. Ich schau mal ob ich das zeitnah hinbekomme.

> > So please revise your PR if possible.
> 
> Ja, da habe ich auch mehr Fragen als Antworten. Wie ändert man einen PR? Änder ich den in meinem Repository oder in deinem? oder beiden? Oder besser löschen und neu anfangen?

Das kann ich aber jetzt nicht auch noch dokumentieren, das ist Git/GitHub common sense ;-) Aber im Grunde musst du doch einfach nur sachen in deinem Branch den du extra für diesen PR generiert hast anpassen, ändern und dann landen die hier automatisch. D.h. einfach aktuell deine *.diff Datei abändern und geändert wieder einchecken und pushen, dann landet das hier und dann kann ich erneut darauf basierende kommentieren, etc.

Reply:
> Daran hatte ich mich orientiert. Das bedeutet aber auch, das mein Code immer ausgeführt wird - ob neue oder geänderte SV. Das ist bei einer neuen SV erstmal unschädlich, da die ja noch in keinem Programm ist. Aber vielleicht wäre es doch besser den Block nur auszuführen, wenn ( system.GetVar("createNew") == 0 ) ? Was meinst Du? Oder gleich hier einfügen?
> 
> [...]
> 
> Wobei mich daran stören würde, das ich die Programme dann vor den SV ändere.

Also ich würde das (suchen von referenzen in Programmen und abändern der typen, etc.) vmtl. in Zeile 851 vor dem `if (oChn)` starten lassen.

> Und eine Frage die ich noch testen muss: wird dieser Code nur beim Ändern des SV Typ ausgeführt oder auch bei Änderungen der Beschreibung, Name, etc? Denn dann hätte das ungewollte Effekte. Dann müsste ich erst noch prüfen, ob sich der Typ überhaupt geändert hat.

Soweit ich das überblicke wird die Funktion immer aufgerufen wenn man irgendetwas an der Systemvariable ändert oder eine neue generiert. Und ja, dann müsstest du natürlich schauen was das für mögliche Seiteneffekte hat wenn man nur die Beschreibung bzw. den Namen ändern möchte, etc.

Reply:
Wenn ich so drüber nachdenke, dann fände ich es schon ganz elegant, wenn wir dann noch ein Popup anzeigen, das "n Programme angepasst wurden und zu überprüfen sind". Denn das die rein technisch dann wieder arbeiten und das die Logik passt sind ja 2 verschiedene Paar Schuhe. Und der unbedarfte DAU übersieht die Tragweite seiner Handlung ggf. nicht. Nun, der klickt evt. auch einfach die Meldung weg. Aber egal. 
Ich weiß nur nicht, wie man so eine Meldung aufpoppen lassen kann. Und dann müssen ja wohl noch Sprachdateien angepasst werden... Ich werde einfach mal einen Zähler mit einbauen. Ein Popup kann man dann ja immer noch reinfrickeln. 
Btw: ich habe auch nachgedacht konkret die Programme aufzulisten, aber zum einen kann so eine Liste ja sehr lang werden und zum anderen kann man sich die besser durch den "Programme" Button generieren lassen.

Reply:
... oder einfach nur ein Popup anzeigen das die SV in Programmen gefunden wurde und daher ggf. diese zu überprüfen sind. Dann könnte man sich die automatische Anpassung sparen und überlässt es dem Nutzer die Anpassungen vorzunehmen?

Reply:
> ALso den Code habe ich fertig. Aber Github macht mich fertig. Die UI ist meilenweit von dem entfernt was mir in 40 Jahren IT begegnet ist.

Vllt. bist du ja auch einfach noch nicht bei Git angekommen. Solltest du git kennen dann ist das ganze eine einfache nullnummer. GitHub != git. Checke das repository via git kommandozeilen client oder GUI client aus und versuche nicht GitHub Desktop oder sowas zu verwenden. Du musst deine weiteren änderungen doch einfach nur in deinen SingleConditions-update-on-SV-Changes branch committen und dann pushen. Und wenn du nicht weisst wie das geht hilft es sicher allgemeine git dokumentation zu lesen.

Reply:
> hilft es sicher allgemeine git dokumentation zu lesen.

Ich habe heute abend mehr GitHub Doku gelesen als gecodet und getestet. Ehrlich, das macht mir keinen Spaß.

Reply:
Wie gesagt, GitHub != git

Reply:
Aber gut, du hast ja nun ein anderen Pull Request gebaut. dann mach ich hier jetzt mal zu. Und wenn du weitere änderungen hast dann bitte einfach auf dem anderen branch via git committen+pushen dann musst du absolut NIX im github tun dafür das die weiteren änderungen da auftauchen...